### PR TITLE
enhancement(components): Add customizable connection retry options for Pulsar client in Pulsar sink

### PR DESCRIPTION
**Pull request by FRosner**
I want to be able to customize the connection retry options for the Pulsar client. This fixes https://github.com/vectordotdev/vector/issues/21242

### CODE REVIEW

File: src/sinks/pulsar/config.rs [hunk 1]
```diff
@@ -171,6 +176,47 @@ pub enum PulsarCompression {
...
 }
 
+#[configurable_component]
+#[configurable(description = "Custom connection retry options configuration for the Pulsar client.")]
+#[derive(Clone, Debug)]
+pub struct CustomConnectionRetryOptions {
+    /// Minimum delay between connection retries.
+    #[configurable(metadata(docs::type_unit = "milliseconds"))]
+    #[configurable(metadata(docs::examples = 10))]
+    pub min_backoff: Duration,
```
Comment (jszwedko) [inline src/sinks/pulsar/config.rs:186]:
```suggestion
    pub min_backoff_ms: Duration,
```

We generally suffix duration types with their unit per https://github.com/vectordotdev/vector/blob/master/docs/specs/configuration.md#option-naming

File: src/sinks/pulsar/config.rs [hunk 2]
```diff
@@ -171,6 +176,47 @@ pub enum PulsarCompression {
...
+
+    /// Maximum delay between reconnection retries.
+    #[configurable(metadata(docs::type_unit = "seconds"))]
+    #[configurable(metadata(docs::examples = 30))]
+    pub max_backoff: Duration,
```
Comment (jszwedko) [inline src/sinks/pulsar/config.rs:191]:
```suggestion
    pub max_backoff_secs: Duration,
```

File: src/sinks/pulsar/config.rs [hunk 3]
```diff
@@ -171,6 +176,47 @@ pub enum PulsarCompression {
...
+
+    /// Maximum number of connection retries.
+    #[configurable(metadata(docs::examples = 12))]
+    pub max_retries: u32,
+
+    /// Time limit to establish a connection.
+    #[configurable(metadata(docs::type_unit = "seconds"))]
+    #[configurable(metadata(docs::examples = 10))]
+    pub connection_timeout: Duration,
```
Comment (jszwedko) [inline src/sinks/pulsar/config.rs:200]:
```suggestion
    pub connection_timeout_secs: Duration,
```

File: src/sinks/pulsar/config.rs [hunk 4]
```diff
@@ -171,6 +176,47 @@ pub enum PulsarCompression {
...
+
+    /// Keep-alive interval for each broker connection.
+    #[configurable(metadata(docs::type_unit = "seconds"))]
+    #[configurable(metadata(docs::examples = 60))]
+    pub keep_alive: Duration,
```
Comment (jszwedko) [inline src/sinks/pulsar/config.rs:205]:
```suggestion
    pub keep_alive_secs: Duration,
```

**Review by jszwedko — COMMENTED** [review]
Apologies for the delay in review @FRosner . I think this looks like a pretty straight-forward addition to me. In addition to my few inline comments, could you:

- Add a changelog entry per https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md ?
- Run `make generate-component-docs` to regenerate the component documentation?

Otherwise I think this looks good.

File: src/sinks/pulsar/config.rs (hunk 4 above)
Comment (FRosner) [reply src/sinks/pulsar/config.rs]: https://github.com/vectordotdev/vector/pull/21245/commits/18c535dfe89382f547d0a700bf38e4eb350f1c39

File: src/sinks/pulsar/config.rs (hunk 3 above)
Comment (FRosner) [reply src/sinks/pulsar/config.rs]: https://github.com/vectordotdev/vector/pull/21245/commits/18c535dfe89382f547d0a700bf38e4eb350f1c39

File: src/sinks/pulsar/config.rs (hunk 2 above)
Comment (FRosner) [reply src/sinks/pulsar/config.rs]: https://github.com/vectordotdev/vector/pull/21245/commits/18c535dfe89382f547d0a700bf38e4eb350f1c39

File: src/sinks/pulsar/config.rs (hunk 1 above)
Comment (FRosner) [reply src/sinks/pulsar/config.rs]: https://github.com/vectordotdev/vector/pull/21245/commits/18c535dfe89382f547d0a700bf38e4eb350f1c39

**Comment by FRosner** [comment]
Thanks @jszwedko! I addressed your feedback. While doing so, I realized that I should probably make those new options optional. I converted both the entire option object to be optional, but also each field. I removed the custom default and am picking the pulsar defaults where no user provided value exists.

I'm completely new to rust, so please let me know if my changes aren't making any sense :)

**Review by aliciascott — APPROVED** [review]
good for docs

**Review by jszwedko — APPROVED** [review]
Thanks @FRosner ! Agreed, these should be optional.

**Comment by FRosner** [comment]
@jszwedko I had to run `cargo fmt` to address the formatting issues. I pushed the changes.

**Pull request merged**